components/renku_data_services/data_connectors/db.py

```diff
@@ -297,6 +300,160 @@ async def delete_data_connector(
+                    data_connector_id}' does not exist or you do not have access to it."
+            )
+
+        project_ids = await self.authz.resources_with_permission(user, user.id, ResourceType.project, Scope.READ)
+
+        async with self.session_maker() as session:
+            stmt = (
+                select(schemas.DataConnectorToProjectLinkORM)
+                .where(schemas.DataConnectorToProjectLinkORM.data_connector_id == data_connector_id)
+                .where(schemas.DataConnectorToProjectLinkORM.project_id.in_(project_ids))
+            )
+            result = await session.scalars(stmt)
+            links_orm = result.all()
+            return [link.dump() for link in links_orm]
+
+    async def get_links_to(
+        self, user: base_models.APIUser, project_id: ULID
+    ) -> list[models.DataConnectorToProjectLink]:
+        """Get links to a given project."""
+        authorized = await self.authz.has_permission(user, ResourceType.project, project_id, Scope.READ)
+        if not authorized:
+            raise errors.MissingResourceError(
+                message=f"Project with id '{
+                    project_id}' does not exist or you do not have access to it."
+            )
+
+        data_connector_ids = await self.authz.resources_with_permission(
+            user, user.id, ResourceType.data_connector, Scope.READ
+        )
+
+        async with self.session_maker() as session:
+            stmt = (
+                select(schemas.DataConnectorToProjectLinkORM)
+                .where(schemas.DataConnectorToProjectLinkORM.project_id == project_id)
+                .where(schemas.DataConnectorToProjectLinkORM.data_connector_id.in_(data_connector_ids))
+            )
+            result = await session.scalars(stmt)
+            links_orm = result.all()
+            return [link.dump() for link in links_orm]
+
+    @with_db_transaction
+    @Authz.authz_change(AuthzOperation.create_link, ResourceType.data_connector)
+    async def insert_link(
+        self,
+        user: base_models.APIUser,
+        link: models.UnsavedDataConnectorToProjectLink,
+        *,
+        session: AsyncSession | None = None,
+    ) -> models.DataConnectorToProjectLink:
+        """Insert a new link from a data connector to a project."""
+        if not session:
+            raise errors.ProgrammingError(message="A database session is required.")
+
+        if user.id is None:
+            raise errors.UnauthorizedError(message="You do not have the required permissions for this operation.")
+
+        data_connector = (
+            await session.scalars(
+                select(schemas.DataConnectorORM).where(schemas.DataConnectorORM.id == link.data_connector_id)
+            )
+        ).one_or_none()
+        if data_connector is None:
+            raise errors.MissingResourceError(
+                message=f"Data connector with id '{link.data_connector_id}' does not exist or you do not have access to it."  # noqa E501
```

Comment:
The function fails before executing if user doesn't have permission. Since we don't check for permission here, the message should just say that the resource is missing.

Reply:
We check for permission in the `@Authz.authz_change(AuthzOperation.create_link, ResourceType.data_connector)` decorator before even entering this function (and we hide client info there as you've mentioned). So, when this is executing, we already know that the user has the permission to access the data connector. Imo, it makes the message a bit confusing (for us when debugging) if we say `you don't have access`. Up to you to decide.